tests/test/subtype.rs

```diff
@@ -121,6 +152,160 @@ fn multi_lifetime() {
+            struct Foo<A> {}
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype(Foo<&'a u32>, Foo<U>),
+                    Subtype(Foo<&'b u32>, Foo<U>)
+                }
+            }
+        } yields {
+            // Result should be opposite multi_lifetime result.
+            "Unique; for<?U1> { substitution [?0 := (&'^0.0 Uint(U32))], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '^0.0: '!1_0 }, \
+            InEnvironment { environment: Env([]), goal: '^0.0: '!1_1 }] }"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within a invariant struct.
+#[test]
+fn multi_lifetime_invariant_struct() {
+    test! {
+        program {
+            #[variance(Invariant)]
+            struct Foo<A> {}
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype(Foo<&'a u32>, Foo<U>),
+                    Subtype(Foo<&'b u32>, Foo<U>)
+                }
+            }
+        } yields {
+            // Because A is invariant, we require the lifetimes to be equal
+            "Unique; substitution [?0 := (&'!1_1 Uint(U32))], lifetime
+            constraints [InEnvironment { environment: Env([]), goal: '!1_0: '!1_1 }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '!1_0 }]"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within slices correctly.
+#[test]
+fn multi_lifetime_slice() {
+    test! {
+        program {
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype([&'a u32], [U]),
+                    Subtype([&'b u32], [U])
+                }
+            }
+        } yields {
+            // Result should be identical to multi_lifetime result.
+            "Unique; for<?U1> { substitution [?0 := (&'^0.0 Uint(U32))], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_0: '^0.0  }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '^0.0  }] }"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within tuples correctly.
+#[test]
+fn multi_lifetime_tuple() {
+    test! {
+        program {
+        }
+
+        goal {
+            forall<'a, 'b> {
+                exists<U> {
+                    Subtype((&'a u32, &'a u32), [U]),
+                    Subtype((&'b u32, &'b u32), [U])
+                }
+            }
+        } yields {
+            // Result should be identical to multi_lifetime result.
+            "Unique; for<?U1> { substitution [?0 := (&'^0.0 Uint(U32))], lifetime constraints [\
+            InEnvironment { environment: Env([]), goal: '!1_0: '^0.0  }, \
+            InEnvironment { environment: Env([]), goal: '!1_1: '^0.0  }] }"
+        }
+    }
+}
+
+/// Tests that the generalizer generalizes lifetimes within tuples correctly.
```

Comment:
```diff
-Tests that the generalizer generalizes lifetimes within tuples correctly.
+Tests that the generalizer generalizes lifetimes within arrays correctly.
```

Reply:
:+1:

Didn't double-check these before pushing - there are a few mistakes in this test and the tuple test which we should push fixes for shortly.